components/service-operator/controllers/serviceaccount.go

```diff
@@ -0,0 +1,178 @@
+	if err := r.KubernetesClient.Get(bg, req.NamespacedName, o); apierrs.IsNotFound(err) {
+		// nothing we can do if the resource has gone missing, so
+		// ignore any not found errors and let the api carry on
+		return controllerutil.OperationResultNone, nil
+	} else if err != nil {
+		// issue communicating with the api
+		// return err and we'll retry later
+		return controllerutil.OperationResultNone, err
+	}
+
+	if _, ok := o.ObjectMeta.Labels[cloudformation.AccessGroupLabel]; !ok {
+		return controllerutil.OperationResultNone, nil
+	}
+
+	// track changes to our object resource and call the main reconcile func
+	var reconcileErr error
+	op, updateErr := controllerutil.CreateOrUpdate(bg, r.KubernetesClient, o, func() error {
+		reconcileErr = r.reconcileObjectWithContext(ctx, req, o)
+		return nil // always try to update
+	})
+	if reconcileErr != nil {
+		return op, reconcileErr
+	} else if updateErr != nil {
+		return op, updateErr
+	}
+	return op, nil
+}
+
+func (r *ServiceAccountController) updatePrincipal(ctx context.Context, o *core.ServiceAccount) (*access.Principal, error) {
+	// List principals with labels set by o.ObjectMeta.Labels
+	list := &access.PrincipalList{}
+	listOptsFunc := func(opts *client.ListOptions) {
+		opts.Namespace = o.GetNamespace()
+		opts.LabelSelector = labels.SelectorFromSet(o.ObjectMeta.Labels)
+	}
+	err := r.KubernetesClient.List(ctx, list, listOptsFunc)
+	if err != nil {
+		return nil, err
+	}
+	principals := list.GetPrincipals()
+	var principal *access.Principal
+	if len(principals) == 0 {
+		// if none are found, make a new one with GenerateName
+		principal = &access.Principal{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: o.GetName(),
+				Namespace:    o.GetNamespace(),
+			},
+		}
+	} else if len(principals) == 1 {
+		// if one is found, use it
+		var ok bool
+		principal, ok = principals[0].(*access.Principal)
+		if !ok {
+			return nil, fmt.Errorf("found principal but could not cast to *access.Principal")
+		}
+	} else if len(principals) > 1 {
+		return nil, fmt.Errorf("multiple principals found with service operator's labels")
+	}
+
+	op, err := controllerutil.CreateOrUpdate(ctx, r.KubernetesClient, principal, func() error {
+		principal.ObjectMeta.Labels = o.ObjectMeta.Labels
+		principal.Spec = access.PrincipalSpec{
+			TrustServiceAccount: o.GetName(),
+		}
+		// mark the principal as owned by the o resource so it gets gc'd
+		if err := controllerutil.SetControllerReference(o, principal, r.Scheme); err != nil {
+			return err
+		}
+		return nil
+	})
+	r.Log.Info("update-principal",
+		"namespace", o.GetNamespace(),
+		"svcacc", o.GetName(),
+		"op", op,
+		"err", err,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return principal, nil
+}
+
+// reconcileObjectWithContext is the main loop, it will mutate "o" with any changes required
+func (r *ServiceAccountController) reconcileObjectWithContext(ctx context.Context, req ctrl.Request, o *core.ServiceAccount) error {
```

Comment:
Can we change the references in this file from `Object` and `o` to `ServiceAccount` (and maybe `s` or `sa`, if we must use shorthands) to make it more readable?

Reply:
```suggestion
func (r *ServiceAccountController) reconcileServiceAccountWithContext(ctx context.Context, req ctrl.Request, serviceAccount *core.ServiceAccount) error {
```

Reply:
note: `gofmt -r 'Object -> ServiceAccount' -w <file>` will do this rename for you